pkg/ipcache/metadata.go

```diff
@@ -224,7 +224,7 @@ func (ipc *IPCache) InjectLabels(ctx context.Context, modifiedPrefixes []netip.P
 			} // else continue below to remove the old entry
 		} else {
 			// Insert to propagate the updated set of labels after removal.
-			newID, _, err = ipc.resolveIdentity(ctx, prefix, prefixInfo, identity.InvalidIdentity)
+			newID, _, err = ipc.resolveIdentity(ctx, prefix, prefixInfo, identity.NumericIdentity(prefixInfo.RestoredIdentity()))
```

Comment:
:eyes: how does this behave if there's another update for this prefix after restore but before the restored identity is removed from the ipcache? I presume this will fill in a numeric identity that is already allocated to a (potentially different) set of labels.

Reply:
`getNextFreeNumericIdentity()` only allocates it if it wasn't already taken :ok_hand:

Reply:
Hmm see next feedback below, I think this last parameter is being ignored currently due to the reservation and lack of unreserve call.

Reply:
Maybe to clarify, my concern here is that we should not reallocate the same number to a different set of labels (or at least not without being super careful about flushing/updating all other users of the numeric identity)

Reply:
Wait, that previous comment didn't quite address what you were getting at.

> Maybe to clarify, my concern here is that we should not reallocate the same number to a different set of labels (or at least not without being super careful about flushing/updating all other users of the numeric identity).

That's a very valid concern, and there's an easy fix for this too: we just need to explicitly Unreserve numeric IDs after the reservation grace period is over. The assumption, then, is that we have reached a quiescent state after the reservation grace period is over, and can safely "free" the pool of reserved identities back in to use.